src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/Utils.java

```diff
@@ -104,34 +105,43 @@ public static long bitsToBytesOrThrow(long bits, Supplier<RuntimeException> exFa
         }
     }
 
-    public static VarHandle makeMemoryAccessVarHandle(Class<?> carrier, boolean skipAlignmentCheck, long alignmentMask, ByteOrder order) {
-        Class<?> baseCarrier = carrier;
-        if (carrier == MemoryAddress.class) {
+    public static VarHandle makeMemoryAccessVarHandle(ValueLayout layout, boolean skipAlignmentCheck) {
+        class VarHandleCache {
+            private static final Map<ValueLayout, VarHandle> handleMap = new ConcurrentHashMap<>();
+            private static final Map<ValueLayout, VarHandle> handleMapNoAlignCheck = new ConcurrentHashMap<>();
```

Comment:
Something to consider later if this is an issue. Since the number of `ValueLayout` instances is fixed, carrier x order = 18, we can use stable arrays with ordinals on the instances.

Reply:
What about alignment?

Reply:
Drat, `skipAlignmentCheck` misled me but perhaps there is still benefit for common constants with 8 bit and size alignment and fallback otherwise.